psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -720,192 +768,882 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+    }
+
+    timer_off("build_JK_matrices()");
+}
 
-            for (int R2 = 0; R2 < nRtask; R2++) {
-                for (int S2 = 0; S2 < nStask; S2++) {
-                    int R = task_shells[R2start + R2];
-                    int S = task_shells[S2start + S2];
-                    int Rsize = primary_->shell(R).nfunction();
-                    int Ssize = primary_->shell(S).nfunction();
+void DirectJK::build_linK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, const std::vector<SharedMatrix>& D,
+                  std::vector<SharedMatrix>& K) {
+
+    if (!lr_symmetric_) {
+        throw PSIEXCEPTION("Non-symmetric K matrix builds are currently not supported in the LinK algorithm.");
+    }
+
+    timer_on("build_linK()");
+
+    // => Zeroing <= //
+
+    for (auto& Kmat : K) {
+        Kmat->zero();
+    }
+
+    // => Sizing <= //
+
+    int nshell = primary_->nshell();
+    int nthread = df_ints_num_threads_;
+
+    // => Atom Blocking <= //
+    // Define the shells of each atom as a task
+    
+    std::vector<int> shell_endpoints_for_atom;
+    std::vector<int> basis_endpoints_for_shell;
+
+    int atomic_ind = -1;
+    for (int P = 0; P < nshell; P++) {
+        if (primary_->shell(P).ncenter() > atomic_ind) {
+            shell_endpoints_for_atom.push_back(P);
+            atomic_ind++;
+        }
+        basis_endpoints_for_shell.push_back(primary_->shell_to_basis_function(P));
+    }
+    shell_endpoints_for_atom.push_back(nshell);
+    basis_endpoints_for_shell.push_back(primary_->nbf());
+
+    // => End Atomic Blocking <= //
+
+    size_t natom = shell_endpoints_for_atom.size() - 1;
+
+    size_t max_functions_per_shell = 0L;
+    for (size_t atom = 0; atom < natom; atom++) {
+        size_t size = 0L;
+        for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+            size += primary_->shell(P).nfunction();
+        }
+        max_functions_per_shell = std::max(max_functions_per_shell, size);
+    }
+
+    if (debug_) {
+        outfile->Printf("  ==> LinK: Atom Blocking <==\n\n");
+        for (size_t atom = 0; atom < natom; atom++) {
+            outfile->Printf("  Atom: %3d, Atom Start: %4d, Atom End: %4d\n", atom, shell_endpoints_for_atom[atom],
+                            shell_endpoints_for_atom[atom + 1]);
+            for (int P = shell_endpoints_for_atom[atom]; P < shell_endpoints_for_atom[atom + 1]; P++) {
+                int size = primary_->shell(P).nfunction();
+                int off = primary_->shell(P).function_index();
+                int off2 = basis_endpoints_for_shell[P];
+                outfile->Printf("    Shell: %4d, Size: %4d, Offset: %4d, Offset2: %4d\n", P, size, off,
+                                off2);
+            }
+        }
+        outfile->Printf("\n");
+    }
+
+    // => Significant Atom Pairs (PQ|-style <= //
+
+    std::vector<std::pair<int, int>> atom_pairs;
```

Comment:
I'm wondering if it would be more efficient/less code smell to use std::vector<std::array<int,2>> instead of std::vector<std::pair<int, int>> as the type for atom pairs. As basically a 2-index tuple, std::pair seems to be better used when the two elements in the pair have different types.

Reply:
I do not think that will significantly impact the performance of our code. What do the core devs think?

Reply:
I expect little impact on performance or code readability. I would have written `std::array<int, 2>` for exactly the reason David said, but I'm not sure it's worth changing, either way.